Use GitHub Actions for CI

phw opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Picard. We
    appreciate your time and interest in helping our project!

    Use this template to help us review your change. Not everything is required,
    depending on your change. Keep or delete what is relevant for your change.
    Remember that it helps us review if you give more helpful info for us to
    understand your change.

    Ensure that you've read through and followed the Contributing Guidelines, in
    [CONTRIBUTING.md](https://github.com/metabrainz/picard/blob/master/CONTRIBUTING.md).
-->

# Summary
Run tests and packaging / deployment with Github Actions
<!--
    Update the checkbox with an [x] for the type of contribution you are making.
-->

* This is a…
    * [ ] Bug fix
    * [ ] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [x] Other
* **Describe this change in 1-2 sentences**:

# Problem

<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->

* JIRA ticket (_optional_): PICARD-XXX
<!--
    Please make sure you prefix your pull request title with 'PICARD-XXX' in order
    for our ticket tracker to link your pull request to the relevant ticket.
-->



# Solution
This performs all our test, build and deployment tasks we have running on travis / appveyor into Github actions.

Advantages:
- All tasks in one place
- Unified and simpler configuration. E.g. the majority of code in `run-tests.yml` is shared between operating systems (there is only a minor difference for installing gettext on Linux and macOS)
- Packaging of the macOS and Windows apps is pretty similar
- We get artifacts storage for all builds (previously only for appveyor)
- It is more flexible, since Github Actions is more than just for CI. E.g. we can react to changes on releases
- It runs faster (full run with all builds and tests is finished in under 10 minutes)

Disadvantages:
- Only newer versions of OS images are available. E.g. there is only a brand new Catalina 10.15. As far as I have tested we should be fine with current builds (macOS builds should run on our current target of macOS >= 10.12, Windows builds work on Windows 7 still). But currently it is not quite clear how Github will deal with updates in the future (e.g. how long they will keep existing versions of images).



<!--
    The details of your change. Talk about technical details, considerations, or
    other interesting points. If you have a lot to say, be more detailed in this
    section.
-->


# Action
Once we merge this I would like to disable the triggers for the appveyor and travisci builds. If needed we could keep the configs around for a while. But reactivating the old builds would be no big deal either way.
<!--
    Other than merging your change, do you want / need us to do anything else
    with your change? This could include reviewing a specific part of your PR.
-->

samj1912 commented:
I am not sure if we want to build on every push. I would rather test on branch pushes and build artifacts only if the tests pass.

samj1912 commented:
Can we also add action ignores for any documentation changes?

samj1912 commented:
I would also cache the dependencies.

phw commented:
> I would also cache the dependencies.

I added caching for pip (see https://github.com/phw/picard/commit/57043c9877518aff00b9d5a120394e689f6d1919), but I reverted it for this PR. For the packages we install there seems to be no real difference, and the caching overhead is at least as large as the marginal savings that might happen. Compare these two builds:

No caching: https://github.com/phw/picard/runs/322548902
Caching with hit: https://github.com/phw/picard/runs/322595415

> Can we also add action ignores for any documentation changes?

I added a list of paths for the workflows. For the macOS / Windows packaging workflows I was strict and added whitelists, this avoid getting those large builds run on unrelated changes. For the test run I added a couple of paths to be ignored. See https://github.com/metabrainz/picard/pull/1374/commits/a4c123bc2593f0a44b7caf4936e3fe0a6468aea3

> I am not sure if we want to build on every push. I would rather test on branch pushes and build artifacts only if the tests pass.

I originally wondered why you had set up the macOS / Windows builds on Appveyor / Travis to happen before the tests. But it turns out this actually was useful in the past, and I actually want to have the builds happen independent on whether tests pass. For one this shows issues with the builds early, so it's easier to address issues in both tests and packaging (instead of having to go through another CI loop). And also being able to run the binaries often is useful even if parts of the test suite fail.

I think with the whitelist approach we limit the number of times the package builds happen, that reduces the overhead for cases where no changes occurred. It makes it a bit more difficult to find the last build packages in some cases (you have to look through previous build commits until you find one which had packages build), but I can live with this.

If the concern is about build times I think we have some options for the macOS builds to become a bit faster. E.g. the brew update takes rather long and does not seem to be necessary. Also notarization polling could happen a bit more frequently. I will experiment with this.

phw commented:
> For the test run I added a couple of paths to be ignored. See [a4c123b](https://github.com/metabrainz/picard/commit/a4c123bc2593f0a44b7caf4936e3fe0a6468aea3)

Maybe we should not add any restrictions when to run the standard test suite. If we want to have the tests required before merging a PR merging will be block if the test have not been run. Also those runs don't take too long.

samj1912 commented on .github/workflows/package-windows.yml:
```diff
@@ -0,0 +1,99 @@
+    - name: Set up Python 3.7
+      uses: actions/setup-python@v1
+      with:
+        python-version: 3.7
```
Any reason we are using 3.7 for packaging?

phw replied:
3.7 is the latest the works with current PyInstaller. They have mostly done the work, see https://github.com/pyinstaller/pyinstaller/issues/4311. So with the next PyInstaller release we should be able to try Python 3.7 for the macOS / Windows builds.

samj1912 commented on .github/workflows/package-windows.yml:
```diff
@@ -0,0 +1,99 @@
+        CODESIGN_PFX_PASSWORD: ${{ secrets.CODESIGN_PFX_PASSWORD }}
+    - name: Archive production artifacts
+      uses: actions/upload-artifact@v1
+      if: always()
```
Should this be always? Or only on success?

phw replied:
Always. We build 3 different packages above, even if the job actually failed I want the created packages to get uploaded.

samj1912 commented on .github/workflows/run-tests.yml:
```diff
@@ -0,0 +1,65 @@
+        flake8 picard --count --show-source --statistics
+        isort --check-only --diff --recursive picard test
+    - name: Test with pytest
+      if: always()
```
Again, this should only run if pip install succeeds. Why is it always run?

phw replied:
That's intentional, I want to see both the results of the syntax checks and the actual test results. If the syntax checks fails the entire job will still be considered failed, but the tests run anyway. Otherwise you get the issues with your commit reported peu à peu.

Actually this option is something I really like about Github Actions, other CI systems often make achieving this behavior difficult.

samj1912 commented on .github/workflows/run-tests.yml:
```diff
@@ -0,0 +1,65 @@
+    - name: Run pip install .
+      run: |
+        python -m pip install --upgrade pip
+        pip install .
```
After installing should we print out the version string just as a test?

phw commented:
I added the version run after pip install and merged this PR so we can get this running for new code changes. We can do improvements to this once we see how well it works.

I also disabled (but not removed) the webhooks for travis and appveyor. But let's keep this configured for a while so we can easily re-enable it if needed.
